// @shopify/eslint-plugin-assemblyscript/lib/index.ts
import rules from "./rules";
import tsplugin from "@typescript-eslint/eslint-plugin";

export = {
  rules: {
    // Alias all of typescript custom rules into assemblyscript rules.
    // This enables pinning the version of the eslint-plugin we are using and remove
    // a peer dependency to @typescript-eslint/eslint-plugin
    ...tsplugin.rules,

    // Our custom set of rules
    ...rules,
  },
  environments: {
    "as-pect": {
      globals: {
        describe: false,
        it: false,
        expect: false,
        log: false,
      },
    },
    builtins: {
      globals: {
        // static checks
        isInteger: false,
        isFloat: false,
        isSigned: false,
        isReference: false,
        isString: false,
        isArray: false,
        isFunction: false,
        isNullable: false,
        isDefined: false,
        isConstant: false,
        isManaged: false,

        // size and alignment
        sizeof: false,
        offsetof: false,
        alignof: false,

        // utility
        assert: false,
        instanciate: false,
        changetype: false,
        idof: false,
        nameof: false,

        // math
        clz: false,
        ctz: false,
        popcnt: false,
        rotl: false,
        rotr: false,
        abs: false,
        max: false,
        min: false,
        ceil: false,
        floor: false,
        copysign: false,
        nearest: false,
        reinterpret: false,
        sqrt: false,
        trunc: false,

        // memory
        store: false,
        load: false,
        memory: false,

        // memory utility
        bswap: false,
        bswap16: false,

        // control flow
        select: false,
        unreachable: false,

        // Keywords
        operator: false,
        unmanaged: false,
        inline: false,

        // stdlib - Constants
        NaN: false,
        Infinity: false,

        // stdlib - Functions
        isNaN: false,
        isFinite: false,
        parseInt: false,
        parseFloat: false,

        // stdlib - Datatypes
        Array: false,
        ArrayBuffer: false,
        DataView: false,
        Date: false,
        Error: false,
        Map: false,
        Math: false,
        Number: false,
        Set: false,
        String: false,
        TypedArray: false,
        I32: false,
        U32: false,
        I64: false,
        U64: false,
        F32: false,
        F64: false,
      },
    },
  },
  configs: {
    recommended: {
      plugins: ["@shopify/eslint-plugin-assemblyscript"],
      extends: ["eslint:recommended"],
      // Use a typescript parser to generate an AST that can be introspected by eslint. This can produce
      // false positive due to the small discrepencies between both languages.
      parser: "@typescript-eslint/parser",
      parserOptions: {
        // Assemblyscript supports import/export like modules.
        sourceType: "module",
        ecmaVersion: 6,
      },
      rules: {
        // Assemblyscript variable declaration is different from typescript.
        // Replaced by assemblyscript/variables rule.
        "no-var": "off",
        "prefer-const": "off",

        // Assemblyscript does not support spread, or rest params. (a.k.a. variadics/splat);
        "prefer-rest-params": "off",
        "prefer-spread": "off",

        // Replaced by their typescript specific implementations.
        "no-array-constructor": "off",
        "no-empty-function": "off",
        "no-unused-vars": "off",
        "no-use-before-define": "off",
        camelcase: "off",

        // Which one do we prefer: `... as Toto` OR `<Toto>...`?
        // Which one do we prefer: `let x: Toto = { ... };` OR `let x = { ... } as T;`
        "@shopify/assemblyscript/consistent-type-assertions": [
          "error",
          { assertionStyle: "angle-bracket" },
        ],

        "@shopify/assemblyscript/ban-types": ["error", {
            "types": {
                "String": {
                    "message": "use 'string' instead",
                    "fixWith": "string",
                },
                "Boolean": {
                    "message": "use 'bool' instead",
                    "fixWith": "bool",
                },
                "undefined": { "message": "not supported." },
                "object": { "message": "not supported." },
            },
            "extendDefaults": false,
        }],

        // Typescript-eslint recommended configuration.
        // ref: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/configs/recommended.json
        "@shopify/assemblyscript/adjacent-overload-signatures": "error",
        "@shopify/assemblyscript/ban-ts-ignore": "off",
        "@shopify/assemblyscript/camelcase": "error",
        "@shopify/assemblyscript/class-name-casing": "error",
        "@shopify/assemblyscript/explicit-function-return-type": "warn",
        "@shopify/assemblyscript/interface-name-prefix": "error",
        "@shopify/assemblyscript/member-delimiter-style": "error",
        "@shopify/assemblyscript/no-array-constructor": "error",
        "@shopify/assemblyscript/no-empty-function": "error",
        "@shopify/assemblyscript/no-empty-interface": "error",
        "@shopify/assemblyscript/no-explicit-any": "warn",
        "@shopify/assemblyscript/no-inferrable-types": "error",
        "@shopify/assemblyscript/no-misused-new": "error",
        "@shopify/assemblyscript/no-non-null-assertion": "off",
        "@shopify/assemblyscript/no-this-alias": "error",
        "@shopify/assemblyscript/no-unused-vars": "warn",
        "@shopify/assemblyscript/no-use-before-define": "off",
        "@shopify/assemblyscript/no-var-requires": "error",
        "@shopify/assemblyscript/prefer-namespace-keyword": "error",
        "@shopify/assemblyscript/triple-slash-reference": "error",
        "@shopify/assemblyscript/type-annotation-spacing": "error",

        // Shopify specific custom rules.
        "@shopify/assemblyscript/no-spread": "error",
        "@shopify/assemblyscript/no-rest-params": "error",
        "@shopify/assemblyscript/no-boolean": "error",
        "@shopify/assemblyscript/no-unsupported-keyword": "error",
        "@shopify/assemblyscript/types": "error",
        "@shopify/assemblyscript/variables": "error",
      },
    },
  },
};
